{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<font size=\"12\">Information Integration Bioinformatics Project</font>\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Settings"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "UNIPROT_URL = 'https://www.uniprot.org/uniprot/?query=cancer%20AND%20reviewed:yes%20AND%20organism:%22Homo%20sapiens%20(Human)%20[9606]%22&format=txt&force=true&sort=score'\n",
    "NCBI_URL = 'ftp://ftp.ncbi.nih.gov/gene/DATA/GENE_INFO/Mammalia'\n",
    "SAVE_XLSX = False\n",
    "DOWNLOAD_FILE = True\n",
    "\n",
    "PSQL_USER = 'adrie'\n",
    "PSQL_PASSWORD = '123'\n",
    "PSQL_HOST = '127.0.0.1'\n",
    "PSQL_PORT = '5432'\n",
    "PSQL_DATABASE = 'adrie'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import pandas as pd\n",
    "import re\n",
    "import requests\n",
    "import urllib.request as request\n",
    "import gzip\n",
    "import psycopg2\n",
    "from psycopg2 import Error"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Uniprot Part"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "if DOWNLOAD_FILE:\n",
    "    url = UNIPROT_URL\n",
    "    r = requests.get(url)  \n",
    "    with open(\"uniprot.txt\",'wb') as f: \n",
    "        f.write(r.content)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "file = open(\"uniprot.txt\", 'r') \n",
    "lines = file.readlines() "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "uniprot = pd.DataFrame(columns=['ID','GN_name','AC','Lenght','Rec_Name','MIM','KW'])\n",
    "uni_synonyms = pd.DataFrame(columns=['ID','GN_name','GN_synonym'])\n",
    "uni_altName = pd.DataFrame(columns=['ID','Alt_name','Short_Alt_name'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "def make_tables(ID, GN_Name, AC, Lenght, Rec_Name, MIM, KW, GN_Synonyms, alts, alt_short, uniprot, uni_synonyms, uni_altName):\n",
    "    #uniprot\n",
    "    uniprot = uniprot.append(pd.DataFrame([[ID, GN_Name, AC, Lenght, Rec_Name, MIM, KW]], \n",
    "                    columns=['ID','GN_name','AC','Lenght','Rec_Name','MIM','KW']), ignore_index=True)\n",
    "    #synonyms\n",
    "    for syn in GN_Synonyms:\n",
    "        uni_synonyms = uni_synonyms.append(pd.DataFrame([[ID, GN_Name, syn]], \n",
    "                    columns=['ID','GN_name','GN_synonym']), ignore_index=True)\n",
    "    #altName\n",
    "    for alt in alts:\n",
    "        short = ''\n",
    "        if alt in alt_short:\n",
    "            short = alt_short[alt]\n",
    "        uni_altName = uni_altName.append(pd.DataFrame([[ID, alt, short]], \n",
    "                    columns=['ID','Alt_name','Short_Alt_name']), ignore_index=True)\n",
    "        \n",
    "    return uniprot, uni_synonyms, uni_altName"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "def ID_handler(line):\n",
    "    sp = line.split()\n",
    "    ID = sp[1]\n",
    "    Lenght = sp[3]\n",
    "    return ID, Lenght\n",
    "\n",
    "def AC_handler(line):\n",
    "    AC = ''\n",
    "    sp = line.split()\n",
    "    for i in range(1,len(sp)):\n",
    "        AC=AC+sp[i]\n",
    "    return AC\n",
    "\n",
    "def DE_handler(line):\n",
    "    result = re.search('=(.*);', line)\n",
    "    return result.group(1)\n",
    "\n",
    "def GN_handler(line, scenario):  # 1.Name=/2.Synonyme=/3.Both\n",
    "    \n",
    "    GN_Name = ''\n",
    "    GN_Synonyms = []\n",
    "    \n",
    "    if scenario == 1:\n",
    "        line = line.split(';')[0]\n",
    "        sp = line[line.find('Name=')+5:].split()\n",
    "        for s in sp:\n",
    "            if ('{' not in s):\n",
    "                if('ECO' not in s):\n",
    "                    GN_Name = s\n",
    "        return GN_Name\n",
    "        \n",
    "        \n",
    "    if scenario == 2:\n",
    "        sys = re.search('Synonyms=(.*)', line)\n",
    "        sp = sys.group(1).split()\n",
    "        \n",
    "        for s in sp:\n",
    "            if ('{' not in s) or ('ECO' not in s):\n",
    "                s = s.replace(',' , '')\n",
    "                s = s.replace(';', '')\n",
    "                GN_Synonyms.append(s)\n",
    "        return GN_Synonyms\n",
    "        \n",
    "        \n",
    "    if scenario == 3:\n",
    "        GN_Name = line[line.find('Name=')+5:line.find(';')]\n",
    "        GN_Name = GN_Name.split()[0]\n",
    "        \n",
    "        sys_line = line[line.find('Synonyms=')+9:]\n",
    "        sys_line = sys_line[:sys_line.find(';')]\n",
    "        sp = sys_line.split()\n",
    "        for s in sp:\n",
    "            s = s.replace(',' , '')\n",
    "            GN_Synonyms.append(s)\n",
    "\n",
    "    return GN_Name, GN_Synonyms\n",
    "\n",
    "def MIM_handler(line): # if gene if MIM\n",
    "    sp = line.split()[2]\n",
    "    MIM = sp.split(';')[0]\n",
    "    return MIM\n",
    "\n",
    "def KW_handler(line): #results should be concated\n",
    "    KW = line.replace('KW   ' , '')\n",
    "    return KW"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "for line in lines: \n",
    "    \n",
    "    if line.startswith('ID'):\n",
    "        KW = ''\n",
    "        ID, Lenght = ID_handler(line)\n",
    "    \n",
    "    if line.startswith('AC'):\n",
    "        AC = AC_handler(line)\n",
    "        \n",
    "    if line.startswith('DE'):\n",
    "        if 'RecName' in line:\n",
    "            alts = []\n",
    "            alt_short = dict()\n",
    "            Rec_Name = DE_handler(line)\n",
    "        if 'AltName' in line:\n",
    "            Alt_name = DE_handler(line)\n",
    "            alts.append(Alt_name)\n",
    "        if 'Short' in line:\n",
    "            alt_short[Alt_name] = DE_handler(line)\n",
    "            \n",
    "    if line.startswith('GN'):\n",
    "        scenario = 0 \n",
    "        GN_Synonyms = []\n",
    "        if (' Name' in line) and ('Synonyms' not in line):\n",
    "            scenario = 1\n",
    "            GN_Name = GN_handler(line, scenario)\n",
    "        if ('Synonyms' in line) and ('Name' not in line):\n",
    "            scenario = 2\n",
    "            GN_Synonyms = GN_handler(line, scenario)\n",
    "        if (' Name' in line) and ('Synonyms' in line):\n",
    "            scenario = 3\n",
    "            GN_Name, GN_Synonyms = GN_handler(line, scenario)\n",
    "\n",
    "    if line.startswith('DR'):\n",
    "        if 'MIM' in line:\n",
    "            if 'gene' in line:\n",
    "                MIM = MIM_handler(line)\n",
    "\n",
    "    if line.startswith('KW'):\n",
    "        KW = KW + KW_handler(line)\n",
    "            \n",
    "    if line.startswith('SQ'): #object finished ==> make tables\n",
    "        uniprot, uni_synonyms, uni_altName = make_tables(ID, GN_Name, AC, Lenght, Rec_Name, MIM, KW, GN_Synonyms, \n",
    "                                                 alts, alt_short, uniprot, uni_synonyms, uni_altName)\n",
    "         "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ID</th>\n",
       "      <th>GN_name</th>\n",
       "      <th>AC</th>\n",
       "      <th>Lenght</th>\n",
       "      <th>Rec_Name</th>\n",
       "      <th>MIM</th>\n",
       "      <th>KW</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>DLEC1_HUMAN</td>\n",
       "      <td>DLEC1</td>\n",
       "      <td>Q9Y238;Q9NSW0;Q9NTG5;</td>\n",
       "      <td>1755</td>\n",
       "      <td>Deleted in lung and esophageal cancer protein 1</td>\n",
       "      <td>604050</td>\n",
       "      <td>Alternative splicing; Cytoplasm; Polymorphism;...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>SDCG8_HUMAN</td>\n",
       "      <td>SDCCAG8</td>\n",
       "      <td>Q86SQ7;O60527;Q3ZCR6;Q8N5F2;Q9P0F1;</td>\n",
       "      <td>713</td>\n",
       "      <td>Serologically defined colon cancer antigen 8</td>\n",
       "      <td>613524</td>\n",
       "      <td>Alternative splicing; Bardet-Biedl syndrome; C...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>KNL1_HUMAN</td>\n",
       "      <td>KNL1</td>\n",
       "      <td>Q8NG31;Q8NHE1;Q8WXA6;Q9HCK2;Q9NR92;</td>\n",
       "      <td>2342</td>\n",
       "      <td>Kinetochore scaffold 1</td>\n",
       "      <td>609173</td>\n",
       "      <td>3D-structure; Alternative splicing; Cell cycle...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>NCOA6_HUMAN</td>\n",
       "      <td>NCOA6</td>\n",
       "      <td>Q14686;A6NLF1;B2RMN5;E1P5P7;Q9NTZ9;Q9UH74;Q9UK86;</td>\n",
       "      <td>2063</td>\n",
       "      <td>Nuclear receptor coactivator 6</td>\n",
       "      <td>605299</td>\n",
       "      <td>Acetylation; Activator; Methylation; Nucleus; ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>LTMD1_HUMAN</td>\n",
       "      <td>LETMD1</td>\n",
       "      <td>Q9BY59;Q9Y3X3;</td>\n",
       "      <td>360</td>\n",
       "      <td>LETM1 domain-containing protein 1</td>\n",
       "      <td>605299</td>\n",
       "      <td>Alternative splicing; Membrane; Mitochondrion;...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "            ID  GN_name                                                 AC  \\\n",
       "0  DLEC1_HUMAN    DLEC1                              Q9Y238;Q9NSW0;Q9NTG5;   \n",
       "1  SDCG8_HUMAN  SDCCAG8                Q86SQ7;O60527;Q3ZCR6;Q8N5F2;Q9P0F1;   \n",
       "2   KNL1_HUMAN     KNL1                Q8NG31;Q8NHE1;Q8WXA6;Q9HCK2;Q9NR92;   \n",
       "3  NCOA6_HUMAN    NCOA6  Q14686;A6NLF1;B2RMN5;E1P5P7;Q9NTZ9;Q9UH74;Q9UK86;   \n",
       "4  LTMD1_HUMAN   LETMD1                                     Q9BY59;Q9Y3X3;   \n",
       "\n",
       "  Lenght                                         Rec_Name     MIM  \\\n",
       "0   1755  Deleted in lung and esophageal cancer protein 1  604050   \n",
       "1    713     Serologically defined colon cancer antigen 8  613524   \n",
       "2   2342                           Kinetochore scaffold 1  609173   \n",
       "3   2063                   Nuclear receptor coactivator 6  605299   \n",
       "4    360                LETM1 domain-containing protein 1  605299   \n",
       "\n",
       "                                                  KW  \n",
       "0  Alternative splicing; Cytoplasm; Polymorphism;...  \n",
       "1  Alternative splicing; Bardet-Biedl syndrome; C...  \n",
       "2  3D-structure; Alternative splicing; Cell cycle...  \n",
       "3  Acetylation; Activator; Methylation; Nucleus; ...  \n",
       "4  Alternative splicing; Membrane; Mitochondrion;...  "
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "uniprot.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ID</th>\n",
       "      <th>GN_name</th>\n",
       "      <th>GN_synonym</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>DLEC1_HUMAN</td>\n",
       "      <td>DLEC1</td>\n",
       "      <td>DLC1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>SDCG8_HUMAN</td>\n",
       "      <td>SDCCAG8</td>\n",
       "      <td>CCCAP</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>SDCG8_HUMAN</td>\n",
       "      <td>SDCCAG8</td>\n",
       "      <td>NPHP10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>KNL1_HUMAN</td>\n",
       "      <td>KNL1</td>\n",
       "      <td>CASC5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>KNL1_HUMAN</td>\n",
       "      <td>KNL1</td>\n",
       "      <td>KIAA1570</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "            ID  GN_name GN_synonym\n",
       "0  DLEC1_HUMAN    DLEC1       DLC1\n",
       "1  SDCG8_HUMAN  SDCCAG8      CCCAP\n",
       "2  SDCG8_HUMAN  SDCCAG8     NPHP10\n",
       "3   KNL1_HUMAN     KNL1      CASC5\n",
       "4   KNL1_HUMAN     KNL1   KIAA1570"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "uni_synonyms.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ID</th>\n",
       "      <th>Alt_name</th>\n",
       "      <th>Short_Alt_name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>DLEC1_HUMAN</td>\n",
       "      <td>Deleted in lung cancer protein 1</td>\n",
       "      <td>DLC-1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>SDCG8_HUMAN</td>\n",
       "      <td>Antigen NY-CO-8</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>SDCG8_HUMAN</td>\n",
       "      <td>Centrosomal colon cancer autoantigen protein</td>\n",
       "      <td>hCCCAP</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>KNL1_HUMAN</td>\n",
       "      <td>ALL1-fused gene from chromosome 15q14 protein</td>\n",
       "      <td>AF15q14</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>KNL1_HUMAN</td>\n",
       "      <td>Bub-linking kinetochore protein</td>\n",
       "      <td>Blinkin</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "            ID                                       Alt_name Short_Alt_name\n",
       "0  DLEC1_HUMAN               Deleted in lung cancer protein 1          DLC-1\n",
       "1  SDCG8_HUMAN                                Antigen NY-CO-8               \n",
       "2  SDCG8_HUMAN   Centrosomal colon cancer autoantigen protein         hCCCAP\n",
       "3   KNL1_HUMAN  ALL1-fused gene from chromosome 15q14 protein        AF15q14\n",
       "4   KNL1_HUMAN                Bub-linking kinetochore protein        Blinkin"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "uni_altName.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "if SAVE_XLSX:\n",
    "    uniprot.to_excel('uniprot_main.xlsx')\n",
    "    uni_synonyms.to_excel('uniprot_synonyms.xlsx')\n",
    "    uni_altName.to_excel('uniprot_altName.xlsx')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# NCBI Part"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "human_genome_file = './Homo_sapiens.gene_info'\n",
    "\n",
    "if DOWNLOAD_FILE:\n",
    "    \n",
    "    baseURL = NCBI_URL\n",
    "    filename = \"Homo_sapiens.gene_info.gz\"\n",
    "\n",
    "    response = request.urlopen(\"/\".join([baseURL, filename, ]))\n",
    "    with open(human_genome_file, \"wb\") as f:\n",
    "        f.write(gzip.decompress(response.read()))\n",
    "        \n",
    "gene_info_file = human_genome_file\n",
    "human_genome_info = pd.read_csv(gene_info_file, delimiter='\\t')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Not cleaned table from dataset"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>#tax_id</th>\n",
       "      <th>GeneID</th>\n",
       "      <th>Symbol</th>\n",
       "      <th>LocusTag</th>\n",
       "      <th>Synonyms</th>\n",
       "      <th>dbXrefs</th>\n",
       "      <th>chromosome</th>\n",
       "      <th>map_location</th>\n",
       "      <th>description</th>\n",
       "      <th>type_of_gene</th>\n",
       "      <th>Symbol_from_nomenclature_authority</th>\n",
       "      <th>Full_name_from_nomenclature_authority</th>\n",
       "      <th>Nomenclature_status</th>\n",
       "      <th>Other_designations</th>\n",
       "      <th>Modification_date</th>\n",
       "      <th>Feature_type</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>9606</td>\n",
       "      <td>1</td>\n",
       "      <td>A1BG</td>\n",
       "      <td>-</td>\n",
       "      <td>A1B|ABG|GAB|HYST2477</td>\n",
       "      <td>MIM:138670|HGNC:HGNC:5|Ensembl:ENSG00000121410</td>\n",
       "      <td>19</td>\n",
       "      <td>19q13.43</td>\n",
       "      <td>alpha-1-B glycoprotein</td>\n",
       "      <td>protein-coding</td>\n",
       "      <td>A1BG</td>\n",
       "      <td>alpha-1-B glycoprotein</td>\n",
       "      <td>O</td>\n",
       "      <td>alpha-1B-glycoprotein|HEL-S-163pA|epididymis s...</td>\n",
       "      <td>20191220</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>9606</td>\n",
       "      <td>2</td>\n",
       "      <td>A2M</td>\n",
       "      <td>-</td>\n",
       "      <td>A2MD|CPAMD5|FWP007|S863-7</td>\n",
       "      <td>MIM:103950|HGNC:HGNC:7|Ensembl:ENSG00000175899</td>\n",
       "      <td>12</td>\n",
       "      <td>12p13.31</td>\n",
       "      <td>alpha-2-macroglobulin</td>\n",
       "      <td>protein-coding</td>\n",
       "      <td>A2M</td>\n",
       "      <td>alpha-2-macroglobulin</td>\n",
       "      <td>O</td>\n",
       "      <td>alpha-2-macroglobulin|C3 and PZP-like alpha-2-...</td>\n",
       "      <td>20200113</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>9606</td>\n",
       "      <td>3</td>\n",
       "      <td>A2MP1</td>\n",
       "      <td>-</td>\n",
       "      <td>A2MP</td>\n",
       "      <td>HGNC:HGNC:8|Ensembl:ENSG00000256069</td>\n",
       "      <td>12</td>\n",
       "      <td>12p13.31</td>\n",
       "      <td>alpha-2-macroglobulin pseudogene 1</td>\n",
       "      <td>pseudo</td>\n",
       "      <td>A2MP1</td>\n",
       "      <td>alpha-2-macroglobulin pseudogene 1</td>\n",
       "      <td>O</td>\n",
       "      <td>pregnancy-zone protein pseudogene</td>\n",
       "      <td>20191221</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>9606</td>\n",
       "      <td>9</td>\n",
       "      <td>NAT1</td>\n",
       "      <td>-</td>\n",
       "      <td>AAC1|MNAT|NAT-1|NATI</td>\n",
       "      <td>MIM:108345|HGNC:HGNC:7645|Ensembl:ENSG00000171428</td>\n",
       "      <td>8</td>\n",
       "      <td>8p22</td>\n",
       "      <td>N-acetyltransferase 1</td>\n",
       "      <td>protein-coding</td>\n",
       "      <td>NAT1</td>\n",
       "      <td>N-acetyltransferase 1</td>\n",
       "      <td>O</td>\n",
       "      <td>arylamine N-acetyltransferase 1|N-acetyltransf...</td>\n",
       "      <td>20200113</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>9606</td>\n",
       "      <td>10</td>\n",
       "      <td>NAT2</td>\n",
       "      <td>-</td>\n",
       "      <td>AAC2|NAT-2|PNAT</td>\n",
       "      <td>MIM:612182|HGNC:HGNC:7646|Ensembl:ENSG00000156006</td>\n",
       "      <td>8</td>\n",
       "      <td>8p22</td>\n",
       "      <td>N-acetyltransferase 2</td>\n",
       "      <td>protein-coding</td>\n",
       "      <td>NAT2</td>\n",
       "      <td>N-acetyltransferase 2</td>\n",
       "      <td>O</td>\n",
       "      <td>arylamine N-acetyltransferase 2|N-acetyltransf...</td>\n",
       "      <td>20200121</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   #tax_id  GeneID Symbol LocusTag                   Synonyms  \\\n",
       "0     9606       1   A1BG        -       A1B|ABG|GAB|HYST2477   \n",
       "1     9606       2    A2M        -  A2MD|CPAMD5|FWP007|S863-7   \n",
       "2     9606       3  A2MP1        -                       A2MP   \n",
       "3     9606       9   NAT1        -       AAC1|MNAT|NAT-1|NATI   \n",
       "4     9606      10   NAT2        -            AAC2|NAT-2|PNAT   \n",
       "\n",
       "                                             dbXrefs chromosome map_location  \\\n",
       "0     MIM:138670|HGNC:HGNC:5|Ensembl:ENSG00000121410         19     19q13.43   \n",
       "1     MIM:103950|HGNC:HGNC:7|Ensembl:ENSG00000175899         12     12p13.31   \n",
       "2                HGNC:HGNC:8|Ensembl:ENSG00000256069         12     12p13.31   \n",
       "3  MIM:108345|HGNC:HGNC:7645|Ensembl:ENSG00000171428          8         8p22   \n",
       "4  MIM:612182|HGNC:HGNC:7646|Ensembl:ENSG00000156006          8         8p22   \n",
       "\n",
       "                          description    type_of_gene  \\\n",
       "0              alpha-1-B glycoprotein  protein-coding   \n",
       "1               alpha-2-macroglobulin  protein-coding   \n",
       "2  alpha-2-macroglobulin pseudogene 1          pseudo   \n",
       "3               N-acetyltransferase 1  protein-coding   \n",
       "4               N-acetyltransferase 2  protein-coding   \n",
       "\n",
       "  Symbol_from_nomenclature_authority Full_name_from_nomenclature_authority  \\\n",
       "0                               A1BG                alpha-1-B glycoprotein   \n",
       "1                                A2M                 alpha-2-macroglobulin   \n",
       "2                              A2MP1    alpha-2-macroglobulin pseudogene 1   \n",
       "3                               NAT1                 N-acetyltransferase 1   \n",
       "4                               NAT2                 N-acetyltransferase 2   \n",
       "\n",
       "  Nomenclature_status                                 Other_designations  \\\n",
       "0                   O  alpha-1B-glycoprotein|HEL-S-163pA|epididymis s...   \n",
       "1                   O  alpha-2-macroglobulin|C3 and PZP-like alpha-2-...   \n",
       "2                   O                  pregnancy-zone protein pseudogene   \n",
       "3                   O  arylamine N-acetyltransferase 1|N-acetyltransf...   \n",
       "4                   O  arylamine N-acetyltransferase 2|N-acetyltransf...   \n",
       "\n",
       "   Modification_date Feature_type  \n",
       "0           20191220            -  \n",
       "1           20200113            -  \n",
       "2           20191221            -  \n",
       "3           20200113            -  \n",
       "4           20200121            -  "
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "human_genome_info.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Needed columns for the project"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>GeneID</th>\n",
       "      <th>Symbol</th>\n",
       "      <th>chromosome</th>\n",
       "      <th>map_location</th>\n",
       "      <th>description</th>\n",
       "      <th>type_of_gene</th>\n",
       "      <th>Symbol_from_nomenclature_authority</th>\n",
       "      <th>Full_name_from_nomenclature_authority</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>A1BG</td>\n",
       "      <td>19</td>\n",
       "      <td>19q13.43</td>\n",
       "      <td>alpha-1-B glycoprotein</td>\n",
       "      <td>protein-coding</td>\n",
       "      <td>A1BG</td>\n",
       "      <td>alpha-1-B glycoprotein</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>2</td>\n",
       "      <td>A2M</td>\n",
       "      <td>12</td>\n",
       "      <td>12p13.31</td>\n",
       "      <td>alpha-2-macroglobulin</td>\n",
       "      <td>protein-coding</td>\n",
       "      <td>A2M</td>\n",
       "      <td>alpha-2-macroglobulin</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>3</td>\n",
       "      <td>A2MP1</td>\n",
       "      <td>12</td>\n",
       "      <td>12p13.31</td>\n",
       "      <td>alpha-2-macroglobulin pseudogene 1</td>\n",
       "      <td>pseudo</td>\n",
       "      <td>A2MP1</td>\n",
       "      <td>alpha-2-macroglobulin pseudogene 1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>9</td>\n",
       "      <td>NAT1</td>\n",
       "      <td>8</td>\n",
       "      <td>8p22</td>\n",
       "      <td>N-acetyltransferase 1</td>\n",
       "      <td>protein-coding</td>\n",
       "      <td>NAT1</td>\n",
       "      <td>N-acetyltransferase 1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>10</td>\n",
       "      <td>NAT2</td>\n",
       "      <td>8</td>\n",
       "      <td>8p22</td>\n",
       "      <td>N-acetyltransferase 2</td>\n",
       "      <td>protein-coding</td>\n",
       "      <td>NAT2</td>\n",
       "      <td>N-acetyltransferase 2</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   GeneID Symbol chromosome map_location                         description  \\\n",
       "0       1   A1BG         19     19q13.43              alpha-1-B glycoprotein   \n",
       "1       2    A2M         12     12p13.31               alpha-2-macroglobulin   \n",
       "2       3  A2MP1         12     12p13.31  alpha-2-macroglobulin pseudogene 1   \n",
       "3       9   NAT1          8         8p22               N-acetyltransferase 1   \n",
       "4      10   NAT2          8         8p22               N-acetyltransferase 2   \n",
       "\n",
       "     type_of_gene Symbol_from_nomenclature_authority  \\\n",
       "0  protein-coding                               A1BG   \n",
       "1  protein-coding                                A2M   \n",
       "2          pseudo                              A2MP1   \n",
       "3  protein-coding                               NAT1   \n",
       "4  protein-coding                               NAT2   \n",
       "\n",
       "  Full_name_from_nomenclature_authority  \n",
       "0                alpha-1-B glycoprotein  \n",
       "1                 alpha-2-macroglobulin  \n",
       "2    alpha-2-macroglobulin pseudogene 1  \n",
       "3                 N-acetyltransferase 1  \n",
       "4                 N-acetyltransferase 2  "
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "columns = ['GeneID', 'Symbol', 'chromosome', 'map_location', 'description', 'type_of_gene', 'Symbol_from_nomenclature_authority', 'Full_name_from_nomenclature_authority']\n",
    "select_df = human_genome_info[columns]\n",
    "select_df.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "As we need only MIM value, we extract it from dbXrefs column"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>GeneID</th>\n",
       "      <th>Symbol</th>\n",
       "      <th>MIM</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>A1BG</td>\n",
       "      <td>138670</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>2</td>\n",
       "      <td>A2M</td>\n",
       "      <td>103950</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>3</td>\n",
       "      <td>A2MP1</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>9</td>\n",
       "      <td>NAT1</td>\n",
       "      <td>108345</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>10</td>\n",
       "      <td>NAT2</td>\n",
       "      <td>612182</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   GeneID Symbol     MIM\n",
       "0       1   A1BG  138670\n",
       "1       2    A2M  103950\n",
       "2       3  A2MP1        \n",
       "3       9   NAT1  108345\n",
       "4      10   NAT2  612182"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "xref_rows = list()\n",
    "\n",
    "for i, series in human_genome_info.iterrows():\n",
    "    gene_id = series.GeneID\n",
    "    symbol = series.Symbol\n",
    "    dbXrefs = series.dbXrefs\n",
    "    if pd.notnull(dbXrefs):\n",
    "        a = dbXrefs.split('|')\n",
    "        if a[0].startswith('MIM:'):\n",
    "            b,c=a[0].split(':', 1)\n",
    "            xref_rows.append((gene_id, symbol, c))\n",
    "        else:\n",
    "            xref_rows.append((gene_id, symbol, ' '))\n",
    "            \n",
    "xref_df = pd.DataFrame(xref_rows, columns=['GeneID','Symbol', 'MIM'])\n",
    "xref_df.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Combine two tables to get needed information from table"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>GeneID_x</th>\n",
       "      <th>GN_Name</th>\n",
       "      <th>chromosome</th>\n",
       "      <th>map_location</th>\n",
       "      <th>description</th>\n",
       "      <th>type_of_gene</th>\n",
       "      <th>Symbol_from_nomenclature_authority</th>\n",
       "      <th>Full_name_from_nomenclature_authority</th>\n",
       "      <th>GeneID_y</th>\n",
       "      <th>MIM</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>A1BG</td>\n",
       "      <td>19</td>\n",
       "      <td>19q13.43</td>\n",
       "      <td>alpha-1-B glycoprotein</td>\n",
       "      <td>protein-coding</td>\n",
       "      <td>A1BG</td>\n",
       "      <td>alpha-1-B glycoprotein</td>\n",
       "      <td>1</td>\n",
       "      <td>138670</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>2</td>\n",
       "      <td>A2M</td>\n",
       "      <td>12</td>\n",
       "      <td>12p13.31</td>\n",
       "      <td>alpha-2-macroglobulin</td>\n",
       "      <td>protein-coding</td>\n",
       "      <td>A2M</td>\n",
       "      <td>alpha-2-macroglobulin</td>\n",
       "      <td>2</td>\n",
       "      <td>103950</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>3</td>\n",
       "      <td>A2MP1</td>\n",
       "      <td>12</td>\n",
       "      <td>12p13.31</td>\n",
       "      <td>alpha-2-macroglobulin pseudogene 1</td>\n",
       "      <td>pseudo</td>\n",
       "      <td>A2MP1</td>\n",
       "      <td>alpha-2-macroglobulin pseudogene 1</td>\n",
       "      <td>3</td>\n",
       "      <td></td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>9</td>\n",
       "      <td>NAT1</td>\n",
       "      <td>8</td>\n",
       "      <td>8p22</td>\n",
       "      <td>N-acetyltransferase 1</td>\n",
       "      <td>protein-coding</td>\n",
       "      <td>NAT1</td>\n",
       "      <td>N-acetyltransferase 1</td>\n",
       "      <td>9</td>\n",
       "      <td>108345</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>10</td>\n",
       "      <td>NAT2</td>\n",
       "      <td>8</td>\n",
       "      <td>8p22</td>\n",
       "      <td>N-acetyltransferase 2</td>\n",
       "      <td>protein-coding</td>\n",
       "      <td>NAT2</td>\n",
       "      <td>N-acetyltransferase 2</td>\n",
       "      <td>10</td>\n",
       "      <td>612182</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   GeneID_x GN_Name chromosome map_location  \\\n",
       "0         1    A1BG         19     19q13.43   \n",
       "1         2     A2M         12     12p13.31   \n",
       "2         3   A2MP1         12     12p13.31   \n",
       "3         9    NAT1          8         8p22   \n",
       "4        10    NAT2          8         8p22   \n",
       "\n",
       "                          description    type_of_gene  \\\n",
       "0              alpha-1-B glycoprotein  protein-coding   \n",
       "1               alpha-2-macroglobulin  protein-coding   \n",
       "2  alpha-2-macroglobulin pseudogene 1          pseudo   \n",
       "3               N-acetyltransferase 1  protein-coding   \n",
       "4               N-acetyltransferase 2  protein-coding   \n",
       "\n",
       "  Symbol_from_nomenclature_authority Full_name_from_nomenclature_authority  \\\n",
       "0                               A1BG                alpha-1-B glycoprotein   \n",
       "1                                A2M                 alpha-2-macroglobulin   \n",
       "2                              A2MP1    alpha-2-macroglobulin pseudogene 1   \n",
       "3                               NAT1                 N-acetyltransferase 1   \n",
       "4                               NAT2                 N-acetyltransferase 2   \n",
       "\n",
       "   GeneID_y     MIM  \n",
       "0         1  138670  \n",
       "1         2  103950  \n",
       "2         3          \n",
       "3         9  108345  \n",
       "4        10  612182  "
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "nbi_main=pd.merge(select_df, xref_df, on='Symbol')\n",
    "nbi_main = nbi_main.rename(columns={\"Symbol\": \"GN_Name\"})\n",
    "nbi_main.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>GeneID</th>\n",
       "      <th>GN_Name</th>\n",
       "      <th>GN_Synonym</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>A1BG</td>\n",
       "      <td>A1B</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>A1BG</td>\n",
       "      <td>ABG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>A1BG</td>\n",
       "      <td>GAB</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>A1BG</td>\n",
       "      <td>HYST2477</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>2</td>\n",
       "      <td>A2M</td>\n",
       "      <td>A2MD</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   GeneID GN_Name GN_Synonym\n",
       "0       1    A1BG        A1B\n",
       "1       1    A1BG        ABG\n",
       "2       1    A1BG        GAB\n",
       "3       1    A1BG   HYST2477\n",
       "4       2     A2M       A2MD"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "synonym_rows = list()\n",
    "for i, series in human_genome_info.iterrows():\n",
    "    gene_id = series.GeneID\n",
    "    symbol = series.Symbol\n",
    "    synonyms = series.Synonyms\n",
    "    if pd.notnull(synonyms):\n",
    "        for synonym in synonyms.split('|'):\n",
    "            synonym_rows.append((gene_id, symbol,synonym))\n",
    "nbi_synonym = pd.DataFrame(synonym_rows, columns=['GeneID', 'GN_Name', 'GN_Synonym'])\n",
    "nbi_synonym.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [],
   "source": [
    "if SAVE_XLSX:\n",
    "    nbi_main.to_excel('NBI_main.xlsx')\n",
    "    xref_df.to_excel('NBI_mim.xlsx')\n",
    "    nbi_synonym.to_excel('NBI_synonym.xlsx')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# SQL Part"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [],
   "source": [
    "def create_tables(user,password,host,port,database):\n",
    "    try:\n",
    "        connection = psycopg2.connect(user = user,\n",
    "                                      password = password,\n",
    "                                      host = host,\n",
    "                                      port = port,\n",
    "                                      database = database)\n",
    "        cursor = connection.cursor()\n",
    "       \n",
    "        query = '''DROP TABLE IF EXISTS Uniprot_Name;\n",
    "                   DROP TABLE IF EXISTS Uniprot_Synonym;\n",
    "                   DROP TABLE IF EXISTS Uniprot;\n",
    "                   DROP TABLE IF EXISTS NBI_Synonym;\n",
    "                   DROP TABLE IF EXISTS NBI;'''\n",
    "        cursor.execute(query)\n",
    "        connection.commit()\n",
    "        \n",
    "            \n",
    "        query = '''\n",
    "        CREATE TABLE Uniprot\n",
    "            (ID TEXT PRIMARY KEY NOT NULL,\n",
    "            GN_name TEXT NOT NULL,\n",
    "            AC TEXT NOT NULL,\n",
    "            Length INT NOT NULL,\n",
    "            Rec_Name TEXT NOT NULL,\n",
    "            MIM TEXT,\n",
    "            KW  TEXT NOT NULL);\n",
    "        CREATE TABLE Uniprot_Name\n",
    "            (ID TEXT NOT NULL,\n",
    "            Alt_name TEXT  NOT NULL,\n",
    "            Short_Alt_name  TEXT);\n",
    "        CREATE TABLE Uniprot_Synonym\n",
    "            (ID TEXT  NOT NULL,\n",
    "            GN_name TEXT NOT NULL,\n",
    "            GN_synonym TEXT NOT NULL);\n",
    "        CREATE TABLE NBI \n",
    "            (ID TEXT PRIMARY KEY NOT NULL,\n",
    "            GN_name TEXT NOT NULL,\n",
    "            Chromosone TEXT NOT NULL,\n",
    "            Map_Location TEXT NOT NULL,\n",
    "            Description TEXT NOT NULL,\n",
    "            GN_Type TEXT NOT NULL,   \n",
    "            NA_Symbol TEXT NOT NULL,  \n",
    "            NA_FullName TEXT NOT NULL,\n",
    "            GN_ID INT NOT NULL,\n",
    "            MIM TEXT);\n",
    "        CREATE TABLE NBI_Synonym\n",
    "            (ID TEXT NOT NULL,\n",
    "            GN_Name TEXT NOT NULL,\n",
    "            GN_Synonym TEXT NOT NULL);'''\n",
    "\n",
    "        cursor.execute(query)\n",
    "        connection.commit()\n",
    "        print(\"Table created successfully in PostgreSQL \")\n",
    "\n",
    "    except (Exception, psycopg2.DatabaseError) as error :\n",
    "        print (\"Error while creating PostgreSQL table\", error)\n",
    "    finally:\n",
    "        #closing database connection.\n",
    "            if(connection):\n",
    "                cursor.close()\n",
    "                connection.close()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [],
   "source": [
    "def column_table(dataframe):\n",
    "    liste = []\n",
    "    for c in dataframe:\n",
    "        liste.append(c)\n",
    "    return liste\n",
    "\n",
    "\n",
    "def update(user,password,host,port,database,table,dataframe):\n",
    "    try:\n",
    "        connection = psycopg2.connect(user = user,\n",
    "                                      password = password,\n",
    "                                      host = host,\n",
    "                                      port = port,\n",
    "                                      database = database)\n",
    "        cursor = connection.cursor()\n",
    "        nb_column = len(dataframe.columns) - 1\n",
    "        column = column_table(dataframe)\n",
    "        for index, row in dataframe.iterrows():\n",
    "            query = \"INSERT INTO \"+table+\" VALUES (\"+\"%s,\"*nb_column+\"%s) ON CONFLICT DO NOTHING\"\n",
    "            cursor = connection.cursor()\n",
    "            data = ()\n",
    "            for c in column:\n",
    "                data +=(row[c],)\n",
    "            cursor.execute(query, data)\n",
    "            connection.commit()\n",
    "        print (\"Records inserted successfully into \"+table+\" table\")\n",
    "    except (Exception, psycopg2.DatabaseError) as error :\n",
    "        print (\"Error while creating PostgreSQL table\", error)\n",
    "    finally:\n",
    "         if(connection):\n",
    "                cursor.close()\n",
    "                connection.close()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [],
   "source": [
    "def select(user,password,host,port,database,query):\n",
    "    try:\n",
    "        connection = psycopg2.connect(user = user,\n",
    "                                      password = password,\n",
    "                                      host = host,\n",
    "                                      port = port,\n",
    "                                      database = database)\n",
    "        cursor = connection.cursor()\n",
    "        cursor.execute(query)\n",
    "        connection.commit()\n",
    "        records = cursor.fetchall()\n",
    "        print(\"Total rows: \", len(records))\n",
    "        print(\"Printing each row\")\n",
    "        for row in records:\n",
    "            result = \"\"\n",
    "            for x in range(len(row)):\n",
    "                result += str(row[x])+\"\\t \"\n",
    "            print(result)\n",
    "        cursor.close()\n",
    "    except (Exception, psycopg2.DatabaseError) as error :\n",
    "        print (\"Error while creating PostgreSQL table\", error)\n",
    "    finally:\n",
    "         if(connection):\n",
    "                cursor.close()\n",
    "                connection.close()\n",
    "                print(\"PostgreSQL connection is closed\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Table created successfully in PostgreSQL \n",
      "Records inserted successfully into uniprot table\n",
      "Records inserted successfully into uniprot_synonym table\n",
      "Records inserted successfully into uniprot_name table\n",
      "Records inserted successfully into NBI table\n",
      "Records inserted successfully into NBI_Synonym table\n"
     ]
    }
   ],
   "source": [
    "create_tables(PSQL_USER, PSQL_PASSWORD, PSQL_HOST, PSQL_PORT, PSQL_DATABASE) \n",
    "update(PSQL_USER, PSQL_PASSWORD, PSQL_HOST, PSQL_PORT, PSQL_DATABASE, \"uniprot\", uniprot)\n",
    "update(PSQL_USER, PSQL_PASSWORD, PSQL_HOST, PSQL_PORT, PSQL_DATABASE, \"uniprot_synonym\", uni_synonyms)\n",
    "update(PSQL_USER, PSQL_PASSWORD, PSQL_HOST, PSQL_PORT, PSQL_DATABASE, \"uniprot_name\", uni_altName)\n",
    "update(PSQL_USER, PSQL_PASSWORD, PSQL_HOST, PSQL_PORT, PSQL_DATABASE, \"NBI\", nbi_main)\n",
    "update(PSQL_USER, PSQL_PASSWORD, PSQL_HOST, PSQL_PORT, PSQL_DATABASE, \"NBI_Synonym\", nbi_synonym)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## SQL queries"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Genes with one common name in NCBI and UniProtKB but with different other names in NCBI and UniProtKB"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "query = \"SELECT DISTINCT U.GN_name FROM Uniprot_Synonym U, NBI_Synonym N WHERE U.GN_NAME=N.GN_NAME AND U.GN_Synonym!=N.GN_Synonym\"\n",
    "select(PSQL_USER, PSQL_PASSWORD, PSQL_HOST, PSQL_PORT, PSQL_DATABASE, query)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Total rows:  4717\n",
    "Printing each row\n",
    "SMAD4\t \n",
    "WWOX\t \n",
    "ADGRF4\t \n",
    "FAM83D\t \n",
    "MLLT1\t \n",
    "PARD6A\t \n",
    "CPSF4\t \n",
    "CLIP1\t \n",
    "MSL1\t \n",
    "ZNF142\t \n",
    "AASDH \t \n",
    ".\t \n",
    ".\t \n",
    "."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Genes with one common name in NCBI and UniProtKB but where the MIM terms differs between NCBI and UniProtKB"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "query = \"SELECT DISTINCT U.GN_name FROM Uniprot U, NBI N WHERE U.GN_NAME=N.GN_NAME AND U.MIM!=N.MIM\"\n",
    "select(PSQL_USER, PSQL_PASSWORD, PSQL_HOST, PSQL_PORT, PSQL_DATABASE, query)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Total rows:  908\n",
    "Printing each row\n",
    "CPXCR1\t \n",
    "WIZ\t \n",
    "MAPK1IP1L\t \n",
    "SH3BP5L\t \n",
    "PRPF38B\t \n",
    "INO80D\t \n",
    "RAB3IL1\t \n",
    "TTC29\t \n",
    "RPS6KL1\t \n",
    "LENG1\t\n",
    ".\t \n",
    ".\t \n",
    "."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### All the alternative names for the gene 'UBE2S_HUMAN'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total rows:  3\n",
      "Printing each row\n",
      "E2EPF\t \n",
      "EPF5\t \n",
      "E2-EPF\t \n",
      "PostgreSQL connection is closed\n"
     ]
    }
   ],
   "source": [
    "query = \"SELECT DISTINCT N.GN_synonym FROM Uniprot U,NBI_Synonym N WHERE U.ID='UBE2S_HUMAN' AND N.GN_Name=U.GN_Name UNION SELECT DISTINCT US.GN_synonym FROM Uniprot U, Uniprot_Synonym US WHERE U.ID='UBE2S_HUMAN' AND US.GN_Name=U.GN_Name\"\n",
    "select(PSQL_USER, PSQL_PASSWORD, PSQL_HOST, PSQL_PORT, PSQL_DATABASE, query)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Genes with common MIM name in NCBI and UniProtKB "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "query = \"SELECT DISTINCT U.GN_name, U.MIM FROM Uniprot U, NBI N WHERE U.MIM=N.MIM UNION SELECT DISTINCT N.GN_name, N.MIM FROM Uniprot U, NBI N WHERE U.MIM=N.MIM\"\n",
    "select(PSQL_USER, PSQL_PASSWORD, PSQL_HOST, PSQL_PORT, PSQL_DATABASE, query)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Total rows:  7738\n",
    "Printing each row\n",
    "AADAC\t 600338\t \n",
    "AAK1\t 616405\t \n",
    "AAMP\t 603488\t \n",
    "AARS\t 601065\t \n",
    "AARS1\t 601065\t \n",
    "AARSD1\t 613212\t \n",
    "AASDH\t 614365\t \n",
    "AASDHPPT\t 607756\t \n",
    "AATF\t 608463\t \n",
    "AATK\t 605276\t\n",
    ".\t \n",
    ".\t \n",
    "."
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
